auth_server/authn/oidc_auth.go

```diff
@@ -0,0 +1,352 @@
+		oauth:      conf,
+	}, nil
+}
+
+/*
+This function will be used by the server if the OIDC auth method is selected. It starts the page for OIDC login or
+requests an access token by using the code given by the OIDC provider.
+*/
+func (ga *OIDCAuth) DoOIDCAuth(rw http.ResponseWriter, req *http.Request) {
+	code := req.URL.Query().Get("code")
+	if code != "" {
+		ga.doOIDCAuthCreateToken(rw, code)
+	} else if req.Method == "GET" {
+		ga.doOIDCAuthPage(rw)
+	} else {
+		http.Error(rw, "Invalid auth request", http.StatusBadRequest)
+	}
+}
+
+/*
+Executes tmpl for the OIDC login page.
+*/
+func (ga *OIDCAuth) doOIDCAuthPage(rw http.ResponseWriter) {
+	if err := ga.tmpl.Execute(rw, struct {
+		AuthEndpoint, RedirectURI, ClientId string
+	}{
+		AuthEndpoint: ga.provider.Endpoint().AuthURL,
+		RedirectURI:  ga.oauth.RedirectURL,
+		ClientId:     ga.oauth.ClientID,
+	}); err != nil {
+		http.Error(rw, fmt.Sprintf("Template error: %s", err), http.StatusInternalServerError)
+	}
+}
+
+/*
+Executes tmplResult for the result of the login process.
+*/
+func (ga *OIDCAuth) doOIDCAuthResultPage(rw http.ResponseWriter, un string, pw string) {
+	if err := ga.tmplResult.Execute(rw, struct {
+		Username, Password, RegistryUrl string
+	}{
+		Username:    un,
+		Password:    pw,
+		RegistryUrl: ga.config.RegistryURL,
+	}); err != nil {
+		http.Error(rw, fmt.Sprintf("Template error: %s", err), http.StatusInternalServerError)
+	}
+}
+
+/*
+Requests an OIDC token by using the code that was provided by the OIDC provider. If it was successfull,
+the access token and refresh token is used to create a new token for the users mail address, which is taken from the ID
+token.
+*/
+func (ga *OIDCAuth) doOIDCAuthCreateToken(rw http.ResponseWriter, code string) {
+
+	tok, err := ga.oauth.Exchange(ga.ctx, code)
+	if err != nil {
+		http.Error(rw, fmt.Sprintf("Error talking to OIDC auth backend: %s", err), http.StatusInternalServerError)
+		return
+	}
+	rawIdTok, ok := tok.Extra("id_token").(string)
+	if !ok {
+		http.Error(rw, "No id_token field in oauth2 token.", http.StatusInternalServerError)
+		return
+	}
+	idTok, err := ga.verifier.Verify(ga.ctx, rawIdTok)
+	if err != nil {
+		http.Error(rw, fmt.Sprintf("Failed to verify ID token: %s", err), http.StatusInternalServerError)
+		return
+	}
+	var prof OIDCProfileResponse
+	if err := idTok.Claims(&prof); err != nil {
+		http.Error(rw, fmt.Sprintf("Failed to get mail information from ID token: %s", err), http.StatusInternalServerError)
+		return
+	}
+	if prof.Email == "" {
+		http.Error(rw, fmt.Sprintf("No mail information given in ID token"), http.StatusInternalServerError)
+		return
+	}
+
+	glog.V(2).Infof("New OIDC auth token for %s (Current time: %s, expiration time: %s)", prof.Email, time.Now().String(), tok.Expiry.String())
+
+	dbVal := &TokenDBValue{
+		TokenType:    tok.TokenType,
+		AccessToken:  tok.AccessToken,
+		RefreshToken: tok.RefreshToken,
+		ValidUntil:   tok.Expiry.Add(time.Duration(-30) * time.Second),
```

Comment:
Why is this negative?

Reply:
Yes, I have oriented myself on the implementation of Google auth. But I admit that it is not the ideal way to do it, since access tokens usually have a short lifetime (couple of minutes) and usually the lifetime of refresh tokens is longer. 
Maybe it would be better to use the lifetime of the refresh token minus x seconds/minutes or to enable to predefine a lifetime of the database tokens